src/workload.rs

```diff
@@ -275,40 +278,79 @@ impl LocalClient {
     }
 }
 
-#[derive(serde::Serialize, Default)]
-/// A WorkloadInformation encapsulates all information about workloads in the mesh
+/// WorkloadInformation wraps WorkloadStore, but is able to additionally request resources on-demand.
+/// It is designed to be cheap to clone.
+#[derive(serde::Serialize, Debug, Clone)]
 pub struct WorkloadInformation {
+    #[serde(flatten)]
+    info: Arc<Mutex<WorkloadStore>>,
+
+    /// demand, if present, is used to request on-demand updates for workloads.
+    #[serde(skip_serializing)]
+    demand: Option<Demander>,
+}
+
+impl WorkloadInformation {
+    pub async fn fetch_workload(&self, addr: &IpAddr) -> Option<Workload> {
+        // Wait for it on-demand, *if* needed
+        debug!("lookup workload for {addr}");
+        match self.find_workload(addr) {
+            None => {
+                if let Some(demand) = &self.demand {
+                    info!("workload not found, sending on-demand request for {addr}");
+                    demand.demand(addr.to_string()).await.recv().await;
+                    debug!("on demand ready: {addr}");
+                    self.find_workload(addr)
+                } else {
+                    None
+                }
+            }
+            wl @ Some(_) => wl,
+        }
+    }
+
+    pub fn find_workload(&self, addr: &IpAddr) -> Option<Workload> {
+        let wi = self.info.lock().unwrap();
+        wi.find_workload(addr).cloned()
+    }
+
+    pub async fn find_upstream(&self, addr: SocketAddr) -> (Upstream, bool) {
+        let _ = self.fetch_workload(&addr.ip()).await;
+        let wi = self.info.lock().unwrap();
+        wi.find_upstream(addr)
+    }
+}
+
+/// A WorkloadStore encapsulates all information about workloads in the mesh
+#[derive(serde::Serialize, Default, Debug)]
+struct WorkloadStore {
     workloads: HashMap<IpAddr, Workload>,
-    name_index: HashMap<String, IpAddr>,
     vips: HashMap<SocketAddr, HashSet<Upstream>>,
 }
 
-impl WorkloadInformation {
+impl WorkloadStore {
     fn insert(&mut self, w: Workload) {
         let wip = w.workload_ip;
-        let wname = w.resource_name();
         self.workloads.insert(wip, w);
-        self.name_index.insert(wname, wip);
     }
 
-    fn remove(&mut self, name: String) {
-        if let Some(ip) = self.name_index.remove(&name) {
-            if let Some(_workload) = self.workloads.remove(&ip) {
-                // TODO: store VIPs in workload so we can remove them
-            } else {
-                panic!(
-                    "removed name {} with ip {}, but was not found in workload map",
-                    name, ip
-                )
+    fn remove(&mut self, ip: String) {
+        use std::str::FromStr;
+        let ip: IpAddr = match IpAddr::from_str(&ip) {
```

Comment:
`let else` syntax just came out.. we're on 1.65, right? 

```suggestion
        let Ok(ip) = IpAddr::from_str(&ip) else {
          error!("...")
          return;
       }
       // use `ip` as needed. 
```

Reply:
I changed some other places but in this case `let else` is not amazing since it won't destruct the `Err(e)` for us